File: src/HolzShots/Input/Actions/CaptureClipboardCommand.vb
Imports HolzShots.Composition.Command
Imports HolzShots.Windows.Forms

Namespace Input.Actions
    <Command("captureClipboard")>
    Public Class CaptureClipboardCommand
        Inherits CapturingCommand

        Public Overrides Async Function Invoke(parameters As IReadOnlyDictionary(Of String, String), settingsContext As HSSettings) As Task
            If parameters Is Nothing Then Throw New ArgumentNullException(NameOf(parameters))
            If settingsContext Is Nothing Then Throw New ArgumentNullException(NameOf(settingsContext))

            Dim image = GetClipboardImage()
            If image Is Nothing Then Return

            Dim shot = Screenshot.FromImage(image, Cursor.Position, ScreenshotSource.Clipboard)
            Await ProcessCapturing(shot, settingsContext).ConfigureAwait(True)
        End Function

        Private Shared Function GetClipboardImage() As Bitmap
            Try
                If Not Clipboard.ContainsImage() Then
                    Return Nothing
                End If

                Return CType(Clipboard.GetImage(), Bitmap)
            Catch ex As Exception When _
                            TypeOf ex Is Runtime.InteropServices.ExternalException _
                            OrElse TypeOf ex Is System.Threading.ThreadStateException
                NotificationManager.RetrievingImageFromClipboardFailed(ex)
                Return Nothing
            End Try
        End Function
    End Class
End Namespace

File: src/HolzShots/Input/Action/CaptureClipboardCommand.cs
using System.Drawing;
using System.Windows.Forms;
using HolzShots.Composition.Command;
using HolzShots.Windows.Forms;

namespace HolzShots.Input.Actions
{
    [Command("captureClipboard")]
    public class CaptureClipboardCommand : CapturingCommand
    {
        protected override async Task InvokeInternal(IReadOnlyDictionary<string, string> parameters, HSSettings settingsContext)
        {
            if (parameters == null)
                throw new ArgumentNullException(nameof(parameters));
            if (settingsContext == null)
                throw new ArgumentNullException(nameof(settingsContext));

            var image = GetClipboardImage();
            if (image == null)
                return;

            var shot = Screenshot.FromImage(image, Cursor.Position, ScreenshotSource.Clipboard);
            await ProcessCapturing(shot, settingsContext).ConfigureAwait(true);
        }

        private static Bitmap? GetClipboardImage()
        {
            try
            {
                return Clipboard.ContainsImage()
                    ? (Bitmap)Clipboard.GetImage()
                    : default;
            }
            catch (Exception ex) when (ex is System.Runtime.InteropServices.ExternalException || ex is System.Threading.ThreadStateException)
            {
                NotificationManager.RetrievingImageFromClipboardFailed(ex);
                return default;
            }
        }
    }
}
